rfc/monorepo.md

```diff
@@ -0,0 +1,557 @@
+
+1. Use the local `deck-kayenta` code during the build process of `deck`, using `yarn`'s workspaces feature
+1. Just integrate `deck-kayenta` as a regular `deck` module, like all the other Lerna subprojects under `packages/*`
+
+### The `spin` CLI
+
+This project is actually the most difficult to integrate into the monorepo, as golang builds tend to be really finicky about their environment.  I think Github Actions should afford us the flexibility to partition that out reasonably, but that requires some investigative work.
+
+There are definitely concrete benefits to including it in the build cycle - it can directly benefit from using adjacent Gate libraries, and it's much more testable when colocated against the APIs it is calling.  Setting up ephemeral CLI testing environments (on every change to Gate, even), is well within GHA's capabilities.
+
+### Halyard
+
+Halyard will also be included in the monorepo.  While it is a tool everyone wants to see replaced, it has no replacement yet.  It consumes many Spinnaker service libraries, and would greatly benefit from composite builds.  
+
+### `spinnaker-monitoring`
+
+The `spinnaker-monitoring` project will not be included in the monorepo.  The project is frozen, and not part of a release build.  The final revision's artifact will continue to be pinned in Halyard and re-tagged as needed for external consumption.
+
+### Integrating Gradle Build Plugins
+
+Gradle composite can also build plugins to be used in the same build, allowing us to also combine [Spinnaker's Gradle build plugin](https://github.com/spinnaker/spinnaker-gradle-project) into the project itself.
+
+This is the only non-Spinnaker-service version that was pinned in Spinnaker Java projects, and including `spinnaker-gradle-project` into the monorepo enables trivial iteration on the Gradle build process.
+
+### Releases and Versioning
+
+Currently, each Java project within Spinnaker maintains its own semver-shaped version designation pinned within various other Spinnaker projects.  This is unnecessary with composite builds, as the project builds as one cohesive unit without any dependency versions at all.
+
+However, containers, libraries, and other artifacts must still be published, and there is a need for a contract with library consumers (such as plugins).
+
+#### Spinnaker Release Versioning and Scope
+
+Spinnaker releases will use a date-based versioning scheme of `<year>.<number>.<patch>`, where `number` represents the number of major releases in that calendar year, and `patch` designates new builds produced for reasons outlined below.  The first numbered release of 2023 would therefore be `2023.0`, followed by `2023.1` and so on.  Patches to numbered releases are appended to the numbered release, e.g. `2023.0.1`.
+
+The initial version of a numbered release is considered to have patch version zero, even if not explicitly stated for brevity.  Unqualified numbered releases refer to the latest patch of that numbered release.
```

Comment:
I think for strict semver systems like Jar's and Debian packages we will need to always publish artifacts tagged as `{major}.{minor}.{patch}`? i.e: always add `.0` patch on the end.

We could re-tag a `<year>.<number>` container image when a new patch comes out if there is appetite for consuming those?
This would look like container images such as alpine where the same container SHA might be tagged: `3.17.1`, `3.17`, `3`, `latest`

Reply:
> Unqualified numbered releases refer to the latest patch of that numbered release.
I don't think we can re-publish jar's and debs so it's only containers where we have the luxury of tags.

Apologies if I have misread.

Reply:
> I think for strict semver systems like Jar's and Debian packages we will need to always publish artifacts tagged as {major}.{minor}.{patch}? i.e: always add .0 patch on the end.

I'm not sure about Debian packages, but jars aren't that strict about their versioning - we mostly just need to be conscious of what Gradle considers the "ordering" to be.  

> I don't think we can re-publish jar's and debs so it's only containers where we have the luxury of tags.

That is true - that line was only meant to refer to images and usage in the general vernacular.  I'll clarify it.  

> We could re-tag a <year>.<number> container image when a new patch comes out if there is appetite for consuming those?

I'm sure there would be - we make a lot of use of `<release>.x`-tagged containers, and I would guess that many others do too.